cli/upload.py

```diff
@@ -326,23 +327,47 @@ def _compose_file_mapping(
                     continue
                 else:
                     missing_required_files.append(source_path)
+            res.append([source_path, f"gs://{upload_info.gcs_bucket}/{gcs_uri}"])
 
-        # gsutil treats brackets in a gs-uri as a character set
-        # see https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames#other-wildcard-characters
-        # this replaces opening with a generic wildcard, which should map to only a single file
         else:
-            if "[" in source_path and "]" in source_path:
-                source_path = source_path.replace("[", "?")
-
-            sub = subprocess.run(["gsutil", "ls", source_path], capture_output=True)
-            if sub.returncode != 0:
-                if source_path.replace("?", "[") in upload_info.optional_files:
+            # separate by bucket, to do single ls per bucket
+            # gsutil treats brackets in a gs-uri as a character set
+            # see https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames#other-wildcard-characters
+            # this replaces opening with a generic wildcard, which can only map to only a single bucket
+            bucket = source_path[5:].split("/")[0].replace("[", "?")
+            if bucket not in gs_uris_to_check:
+                gs_uris_to_check[bucket] = {}
+            gs_uris_to_check[bucket][source_path] = gcs_uri
+
+    print(gs_uris_to_check)
+
+    # smart chcek of gs:// URIs
+    # separate by bucket, to do single ls per bucket
+    # then check in the return for all the files
+    for bucket, gs_file_mapping in gs_uris_to_check.items():
+        sub = subprocess.run(["gsutil", "ls", "-r", bucket], capture_output=True)
+        if sub.returncode != 0:
+            raise Exception(
+                f"Error getting {bucket} to check files: {sub.stderr.decode('utf-8'), }"
+            )
+        # didn't fail
+
+        # return in the format
+        # folder:
+        # gs://bucket/[file]
+        # gs://bucket/[file...]
+        #
+        # [folder...]:
+        file_list = [f.strip() for f in sub.stdout.decode("utf-8").split("\n")]
+        file_list = [f for f in file_list if f and not f.endswith(":")]
```

Comment:
A couple nitpicks: it looks like `file_list` is only used for checking membership below, so it's better to make it a `set`. Moreover, rather than doing two list comprehension passes, I think this would be more readable as a single for-loop:
```python
filenames = set()
for filename in sub.stdout.decode("utf-8").split("\n"):
    cleaned_filename = filename.strip()
    if cleaned_filename != "" and not cleaned_filename.endswith(":"):
        filenames.add(cleaned_filename)

Reply:
in next commit